woocommerce/woocommerce
# WooCommerce Blocks: Add a GitHub Action to create issues for flaky E2E tests

### Changes proposed in this Pull Request:





This PR introduces a Github Action, which will create GitHub issues for flaky (WooCommerce Blocks) E2E tests. I used `@wordpress/report-flaky-tests` package from Gutenberg. Basically, we are replicating the logic which is used by Gutenberg to report flaky tests. You may check out [this code](https://github.com/WordPress/gutenberg/blob/13717c6d83b3548fc967f1006cffb09093e1faec/.github/workflows/end2end-test.yml) in Gutenberg repo if you wanna take a look at their code. 
- It will create GitHub issues for flaky tests when PR is merged into the trunk.
	- **Note:** It's not possible to create issues from PRs because of [this if condition](https://github.com/WordPress/gutenberg/blob/13717c6d83b3548fc967f1006cffb09093e1faec/packages/report-flaky-tests/src/run.ts#L140-L144) in `@wordpress/report-flaky-tests` package. Also, It makes sense, because we don't wanna run it whenever a new commit is pushed on a PR. 

For testing & implementation, I created a [copy of WooCommerce repo](https://github.com/imanish003/woocommerce-copy) to avoid creating flaky tests issues on main repo. You can see sample flaky issues [here](https://github.com/imanish003/woocommerce-copy/issues) which were generated by this new Github Action. 

As I explained above, if this Github Action is run by PR then it won't work. Therefore, another way to test it is by manually running the Github action. For testing, I ran it manually as you can see [here](https://github.com/woocommerce/woocommerce/actions/runs/9287897985/job/25560621326) which created following flaky tests issues:
- [[Flaky Test] should show only products that match the filter #47970](https://github.com/woocommerce/woocommerce/issues/47970)
- [[Flaky Test] Enable company field #47968](https://github.com/woocommerce/woocommerce/issues/47968)

In the testing steps, I will explain more in detail about "how it can be run manually using GitHub UI."

Implementing this feature will help us track and address flaky tests more efficiently, ensuring more stable and reliable test suites.

Closes #47047 



### How to test the changes in this Pull Request:

> [!NOTE]
> To be tested by developers only. 



1. Create a new branch from this branch (`add/47047-blocks-e2e-implement-flaky-test-reporter`)
2. Add new E2E tests file: `plugins/woocommerce-blocks/tests/e2e/tests/demo.spec.ts`
3. Add some flaky tests to it. Here is what you can use:
```TS
/**
 * External dependencies
 */
import { test, expect } from '@woocommerce/e2e-playwright-utils';

test( 'Passing test', async ( { page } ) => {
	await page.goto( 'https://example.com' );
	expect( true ).toBeTruthy();
} );

test( 'Flaky test 1010', async ( { page }, testInfo ) => {
	await page.goto( 'https://example.com' );
	// Introduce flakiness
	if ( testInfo.retry < 2 ) {
		throw new Error( 'Flaky failure' );
	}
	expect( true ).toBeTruthy();
} );

test( 'Flaky test 1011', async ( { page }, testInfo ) => {
	await page.goto( 'https://example.com' );
	// Introduce flakiness
	if ( testInfo.retry < 2 ) {
		throw new Error( 'Flaky failure' );
	}
	expect( true ).toBeTruthy();
} );

test( 'Always failing test', async ( { page } ) => {
	await page.goto( 'https://example.com' );
	expect( false ).toBeTruthy();
} );

```

4. Push the branch to remote
5. Go to ["Blocks Playwright Tests" action](https://github.com/woocommerce/woocommerce/actions/workflows/blocks-playwright.yml)
6. Click on "Run workflow" -> Select your branch -> Click on Run workflow (See screenshot below)
	- **Note:** Because of [this if condition](https://github.com/WordPress/gutenberg/blob/13717c6d83b3548fc967f1006cffb09093e1faec/packages/report-flaky-tests/src/run.ts#L140-L144), we are running this Action manually. 
![image](https://github.com/woocommerce/woocommerce/assets/16707866/f922abdf-22d7-4dda-989c-8c4dac9815a9)


7. Wait for the Job to run successfully. After that, verify that you can see new GitHub issues [here](https://github.com/woocommerce/woocommerce/issues) for flaky tests you added to your branch. 
   - ⚠️ Please close the created issues for flaky tests in `demo.spec.ts` after testing. 



### Changelog entry




<details>

#### Significance




#### Type




#### Message 
WooCommerce Blocks: Added a GitHub Action to create issues for flaky E2E tests

#### Comment 

</details>
